File: src/main/java/me/testaccount666/serversystem/commands/executables/back/CommandBack.java
package me.testaccount666.serversystem.commands.executables.back;

import me.testaccount666.serversystem.commands.ServerSystemCommand;
import me.testaccount666.serversystem.commands.executables.AbstractServerSystemCommand;
import me.testaccount666.serversystem.managers.PermissionManager;
import me.testaccount666.serversystem.managers.globaldata.MappingsData;
import me.testaccount666.serversystem.userdata.ConsoleUser;
import me.testaccount666.serversystem.userdata.User;
import org.bukkit.command.Command;
import org.bukkit.entity.Player;

import static me.testaccount666.serversystem.utils.MessageBuilder.command;
import static me.testaccount666.serversystem.utils.MessageBuilder.general;

@ServerSystemCommand(name = "back")
public class CommandBack extends AbstractServerSystemCommand {

    @Override
    public void execute(User commandSender, Command command, String label, String... arguments) {
        if (!checkBasePermission(commandSender, "Back.Use")) return;

        if (commandSender instanceof ConsoleUser) {
            general("NotPlayer", commandSender).build();
            return;
        }

        var backType = commandSender.getLastBackType();
        backType = backType == BackType.NONE? BackType.DEATH : backType;

        if (arguments.length > 0) backType = switch (arguments[0].toLowerCase()) {
            case "death" -> BackType.DEATH;
            case "teleport", "tp" -> BackType.TELEPORT;
            default -> {
                general("InvalidArguments", commandSender).label(label).build();
                yield null;
            }
        };

        if (backType == null) return;

        var backLocation = switch (backType) {
            case DEATH -> commandSender.getLastDeathLocation();
            case TELEPORT -> commandSender.getLastTeleportLocation();
            default -> {
                general("ErrorOccurred", commandSender).label(label).build();
                throw new IllegalStateException("Unexpected value: ${backType}");
            }
        };

        if (backLocation == null) {
            command("Back.NoBackLocation", commandSender).build();
            return;
        }

        commandSender.getPlayer().teleport(backLocation);

        var finalBackType = backType;
        var typeNameOptional = MappingsData.backType().getBackTypeName(finalBackType);
        var typeName = typeNameOptional.orElse("ERROR");

        command("Back.Success", commandSender)
                .postModifier(message -> message.replace("<TYPE>", typeName)).build();
    }

    @Override
    public boolean hasCommandAccess(Player player, Command command) {
        return PermissionManager.hasCommandPermission(player, "Back.Use", false);
    }

    public enum BackType {
        DEATH,
        TELEPORT,
        NONE
    }
}

File: src/main/kotlin/me/testaccount666/serversystem/commands/executables/back/CommandBack.kt
package me.testaccount666.serversystem.commands.executables.back

import me.testaccount666.serversystem.commands.ServerSystemCommand
import me.testaccount666.serversystem.commands.SimpleCompletion
import me.testaccount666.serversystem.commands.executables.AbstractServerSystemCommand
import me.testaccount666.serversystem.extensions.commandMsg
import me.testaccount666.serversystem.extensions.generalMsg
import me.testaccount666.serversystem.managers.messages.MappingsData.Companion.backType
import me.testaccount666.serversystem.userdata.User
import org.bukkit.command.Command

@ServerSystemCommand(
    "back", simpleCompletions = [
        SimpleCompletion(0, ["teleport", "death"])
    ]
)
class CommandBack : AbstractServerSystemCommand() {
    override fun getUsagePermission(command: Command) = "Back.Use"
    override fun getSyntaxPath(command: Command?) = "Back"

    override fun execute(commandSender: User, command: Command, label: String, vararg arguments: String) {
        if (!isPlayer(commandSender)) return

        val backType = arguments.firstOrNull()?.lowercase()?.let {
            when (it) {
                "death" -> BackType.DEATH
                "teleport", "tp" -> BackType.TELEPORT
                else -> {
                    commandSender.generalMsg("InvalidArguments") {
                        syntax(getSyntaxPath(command))
                        label(label)
                    }
                    BackType.NONE
                }
            }
        } ?: (commandSender.lastBackType.takeUnless { it == BackType.NONE } ?: BackType.DEATH)
        if (backType == BackType.NONE) return

        val backLocation = when (backType) {
            BackType.DEATH -> commandSender.lastDeathLocation
            BackType.TELEPORT -> commandSender.lastTeleportLocation
        } ?: run {
            commandSender.commandMsg("Back.NoBackLocation")
            return
        }

        commandSender.getPlayer()?.teleport(backLocation)

        val typeName = backType(commandSender).getBackTypeName(backType) ?: "ERROR"
        commandSender.commandMsg("Back.Success") {
            postModifier { it.replace("<TYPE>", typeName) }
        }
    }

    enum class BackType {
        DEATH,
        TELEPORT,
        NONE
    }
}
